### ISSUE TITLE
Add support for source location builtins (clang, gcc)

### ISSUE THREAD

**Issue by gbbosak** [microsoft/vscode-cpptools#5360]
Both clang and GCC support these source location builtins. However; VSCode's IntelliSense doesn't recognize them. This prevents code completion/IntelliSense from working properly with code that uses these.

See https://clang.llvm.org/docs/LanguageExtensions.html#source-location-builtins.

**Comment by sean-mcmanus** [comment]
It's working for me. Can you provide more repro details, such as the results of running C/C++: Log Diagnostcs from the Command Palette? It sounds like you may not have your compilerPath and/or intelliSenseMode configured correctly.

![image](https://user-images.githubusercontent.com/19859882/80267735-64d24500-8657-11ea-89e1-dcb3669aad1d.png)

**Comment by gbbosak** [comment]
```
-------- Diagnostics - 4/26/2020, 9:51:12 AM
Version: 0.27.0
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/**"
    ],
    "defines": [],
    "compilerPath": "/mntpnt/fuchsia/prebuilt/third_party/clang/linux-x64/bin/clang++",
    "cStandard": "c11",
    "cppStandard": "c++17",
    "intelliSenseMode": "clang-x64",
    "compileCommands": "${workspaceFolder}/out/default/compile_commands.json",
    "compilerArgs": [],
    "browse": {
        "path": [
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ /mntpnt/fuchsia/src/camera/drivers/usb_video/uvc_format.cc ]:
    /mntpnt/fuchsia/src/camera/drivers/usb_video/uvc_format.cc
Translation Unit Configurations:
[ /mntpnt/fuchsia/src/camera/drivers/usb_video/uvc_format.cc ]:
    Process ID: 11076
    Memory Usage: 123 MB
    Compiler Path: /mntpnt/fuchsia/prebuilt/third_party/clang/linux-x64/bin/clang++
    Includes:
        /mntpnt/fuchsia
        /mntpnt/fuchsia/out/default/gen
        /mntpnt/fuchsia/out/default/fidling/gen/sdk/fidl/fuchsia.camera
        /mntpnt/fuchsia/out/default/fidling/gen/zircon/system/fidl/fuchsia-sysmem
        /mntpnt/fuchsia/sdk
        /mntpnt/fuchsia/out/default/gen/sdk
        /mntpnt/fuchsia/zircon/system/ulib/fit/include
        /mntpnt/fuchsia/zircon/system/ulib/fidl-async/include
        /mntpnt/fuchsia/zircon/system/ulib/zx/include
        /mntpnt/fuchsia/zircon/system/ulib/async/include
        /mntpnt/fuchsia/out/default/fidling/gen/sdk/fidl/fuchsia.hardware.camera
        /mntpnt/fuchsia/out/default/fidling/gen/sdk/fidl/fuchsia.hardware.i2c
        /mntpnt/fuchsia/zircon/system/ulib/fidl/include
        /mntpnt/fuchsia/zircon/system/ulib/sync/include
        /mntpnt/fuchsia/zircon/system/ulib/ddk/include
        /mntpnt/fuchsia/out/default/fidling/gen/zircon/system/fidl/fuchsia-hardware-nand
        /mntpnt/fuchsia/out/default/fidling/gen/zircon/system/fidl/fuchsia-device
        /mntpnt/fuchsia/out/default/fidling/gen/zircon/system/fidl/fuchsia-device-manager
        /mntpnt/fuchsia/out/default/fidling/gen/zircon/system/fidl/fuchsia-hardware-usb-peripheral
        /mntpnt/fuchsia/zircon/system/ulib/ddktl/include
        /mntpnt/fuchsia/zircon/system/ulib/fbl/include
        /mntpnt/fuchsia/zircon/system/dev/lib/device-protocol-i2c/include
        /mntpnt/fuchsia/zircon/system/ulib/zxtest/include
        /mntpnt/fuchsia/out/default/banjoing/gen
        /mntpnt/fuchsia/zircon/system/dev/lib/fake_ddk/include
        /mntpnt/fuchsia/zircon/system/ulib/async-loop/include
        /mntpnt/fuchsia/zircon/system/dev/lib/usb/include
        /mntpnt/fuchsia/zircon/system/dev/lib/dev-operation/include
        /mntpnt/fuchsia/zircon/system/ulib/fzl/include
        /mntpnt/fuchsia/prebuilt/third_party/clang/linux-x64/include/c++/v1
        /usr/local/include
        /mntpnt/fuchsia/prebuilt/third_party/clang/linux-x64/lib/clang/11.0.0/include
        /usr/include/x86_64-linux-gnu
        /usr/include
    Defines:
        _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS
        NDEBUG=1
        _LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS
        _ALL_SOURCE
        ZX_ASSERT_LEVEL=2
        WITH_FRAME_POINTERS=1
    Standard Version: c++17
    IntelliSense Mode: clang-x64
    Other Flags:
        --clang
        --clang_version=110000
Total Memory Usage: 123 MB

```

**Comment by gbbosak** [comment]
@sean-mcmanus This works when VSCode is in GCC intellisense mode but fails when it's in Clang mode. Clang has the same methods as GCC here.

**Comment by gbbosak** [comment]
Additionally; here is a c_cpp_properties.json file that can be used to repro this issue
```
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/clang",
            "cStandard": "c11",
            "cppStandard": "c++17",
            "intelliSenseMode": "clang-x64"
        }
    ],
    "version": 4
}
```

**Comment by sean-mcmanus** [comment]
Okay. Yeah, I repro the issue with clang, but not gcc. The issue is inherited from VS, so I filed a bug on it, but the bug hasn't shown up yet...

**Comment by sean-mcmanus** [comment]
...it's at https://developercommunity.visualstudio.com/content/problem/1009195/cc-intellisense-linux-mode-doesnt-recognize-builti.html

**Comment by androm3da** [comment]
@sean-mcmanus, would target-specific builtins be inhibited by this same issue?  Or are these independent of each other?

Also: can we contribute a fix in this area?

**Comment by sean-mcmanus** [comment]
I'm not sure what target-specific builtins are, but our parser may not understand them. Our parser is closed source.

**Comment by androm3da** [comment]
> I'm not sure what target-specific builtins are, but our parser may not understand them. Our parser is closed source.

From this issue `__builtin_FILE()`, `__builtin_FUNCTION()`, etc are target-independent builtins.

But these ones shown below are target-specific ones.  It would seem to me that maybe we need a different `clang-xyz` intelliSenseMode option for each?

* https://clang.llvm.org/docs/LanguageExtensions.html#low-level-arm-exclusive-memory-builtins

```
T __builtin_arm_ldrex(const volatile T *addr);
T __builtin_arm_ldaex(const volatile T *addr);
int __builtin_arm_strex(T val, volatile T *addr);
int __builtin_arm_stlex(T val, volatile T *addr);
void __builtin_arm_clrex(void);
```

* https://gcc.gnu.org/onlinedocs/gcc/Basic-PowerPC-Built-in-Functions-Available-on-all-Configurations.html#Basic-PowerPC-Built-in-Functions-Available-on-all-Configurations

```
void __builtin_mtfsf (const int, double);
void __builtin_mtfsb0 (const int);
void __builtin_mtfsb1 (const int);
```

**Comment by sean-mcmanus** [comment]
This got fixed a year or so ago (around version 1.0.1 or later).
